Add REPL-completion for keyword arguments

Liozou opened the pull request:
Closes #43437 by implementing the feature.

## The feature

I think it's a convenient non-disruptive feature to have, thank you @ronisbr for thinking of it! The behaviour in this branch:
```julia
julia> function test(; key1, key2, key3)
          return nothing
       end
test (generic function with 1 method)

julia> test(; key#TAB
key1=    key2=    key3=    keys     keytype
```
As you can see, all suggested keyword arguments have a finishing `=`, both because it's useful to be able to distinguish them, and because that's probably what you want anyway (feel free to comment if you disagree!). If you want to use the `foo(; bar)` syntax, meaning `foo(; bar=bar)`, then it's fine as well since for that you need `bar` to exist, so it will be part of the suggestions:
```julia
julia> key1 = 3;

julia> test(; key#TAB
key1     key1=    key2=    key3=    keys     keytype
```

## The intentionally not-implemented

In the cases like the previous example, I thought about restricting the REPL suggestions to only those keyword arguments, since any argument following a semicolon or another keyword argument must itself be a keyword argument, and thus, be of the form `bar=barvalue` or just `bar` if `bar` is both a known name and a keyword argument. But there is the special case of splat keyword arguments, of the form `foo(; kwargs...)` which is an exception since `kwargs` can be about anything. So I kept all the usual names for suggestions, I thought I'd mention the reasoning here for reference.

## Update

See https://github.com/JuliaLang/julia/pull/43536#issuecomment-1037631704 below for a slightly more detailed view on what this PR implements.

(EDIT: I removed parts of this message which were not relevant anymore, mostly thanks to the PRs referenced below and some updates of this PR.)

aviatesk commented:
It's a bit tricky to follow logic by code review, so I played with this PR myself. What I found so far is:

```julia
julia> foo(a; foo=0) = println(a, foo);
julia> obj = (; foo = 42);
julia> foo(1; obj.| # => completes to `foo(1; foo`, which doesn't seem intended
```

Also, how about suggesting keyword completions even if we haven't pressed any preceding characters?
E.g., on top of the above case, we may want to have
```julia
julia> foo(1; | # => suggests `key=`
```
?

Liozou commented:
Thank you very much for spotting this! I had accidentally optimized away the special-casing for character `.`, which is a bit special... It's now fixed and tested, along with kwargs containing `!`.

Regarding the second part of your comment, I had thought about it but initially decided against, because TAB-completion at this point already yields something when the function has multiple methods:
```julia
julia> extrema(a; #TAB
extrema(r::AbstractRange) in Base at range.jl:865                       extrema(s::BitSet) in Base at bitset.jl:435
extrema(x::Real) in Base at operators.jl:556                            extrema(f, x::Real) in Base at operators.jl:557
extrema(A::AbstractArray; dims) in Base at multidimensional.jl:1737     extrema(f, A::AbstractArray; dims) in Base at multidimensional.jl:1748
julia> extrema(a, b; #TAB
extrema(f, x::Real) in Base at operators.jl:557                         extrema(f, A::AbstractArray; dims) in Base at multidimensional.jl:1748
```

But I agree that this could be improved as well on several fronts, in particular:
- TAB-completion at this point should only show methods compatible with the signature of the function so far. For instance, `extrema(f, A::AbstractArray; dims) in Base at multidimensional.jl:1748` should not appear after `extrema(a; #TAB` since the semicolon indicates that there won't be a second positional argument
- likewise, there should not be methods that do not accept keyword arguments like `extrema(f, x::Real) in Base at operators.jl:557`

And more related to this current PR, there is the question of whether this kind of TAB-completion should appear even if there is only one compatible method. If not, as is currently the case, then I can add the keyword suggestions you propose, since nothing else appears anyway (like in the function you show above). Otherwise, if the method signature appears, then the names of the possible keywords will be shown as well, so there is no need to propose all keyword autocompletions (it would look messy if both are present).
What do you think?

Liozou commented:
I forgot to mention that the kind of TAB-completion above depends on the inferrability of the arguments so far (which is good!), for instance:
```julia
julia> extrema(f, a; #TAB
extrema(f, x::Real) in Base at operators.jl:557                         extrema(f, A::AbstractArray; dims) in Base at multidimensional.jl:1748
julia> extrema(x->x, [1]; #TAB
extrema(f, A::AbstractArray; dims) in Base at multidimensional.jl:1748
```
and I was wrong about the idea that it depends on the number of methods of the function. It seems to be something else, consider:
```julia
julia> myfunc() = 0
myfunc (generic function with 1 method)

julia> myfunc(a) = a
myfunc (generic function with 2 methods)

julia> myfunc(a, b) = b
myfunc (generic function with 3 methods)

julia> myfunc( # TAB  -> nothing
julia> myfunc(a, b::Int) = a
myfunc (generic function with 4 methods)

julia> myfunc( #TAB -> why only the last one? Because of the type annotation?
myfunc(a, b::Int64) in Main at REPL[89]:1
```

So anyway, I think this whole thing is probably worth another PR.

ronisbr commented:
I just want to point out that the feature implemented by this PR cover 100% my initial request! Thanks @Liozou for the implementation :)

Liozou commented:
Rebased on top of #43577 to fix a conflicting test

Liozou commented:
With https://github.com/JuliaLang/julia/pull/43577 and https://github.com/JuliaLang/julia/pull/43865 merged, most of the previous conversation is moot, and I updated the PR since. So here is the current (hopefully final) version of this PR:

### Commits

- Remove method completion suggestion when the function call has incompatible keywords. For instance, `extrema(3; foobar=5, #TAB` does not suggest anything anymore, since `foobar` is not a valid kwarg for any `extrema` method accepting an `Int` as first argument.
- Add kwarg autocompletion. That's already explained in the initial PR message.
- When completing a method call (or a kwarg of a method call) after using a semicolon, do not provide suggestions of methods that do not have an appropriate number of arguments. For instance, `extrema(; # TAB` does not suggest anything anymore since there is no valid `extrema` call with no positional argument.

All three main changes are implemented in three separate commits, for easier review. These also include tests.

### Example

An example to illustrate some of these changes. it up, with the current PR:
```julia
julia> foo(a::Int, b, args...; kwarg) = nothing;

julia> foo(3, 4, 5, 6, kw #TAB
julia> foo(3, 4, 5, 6, kwarg=   # (replaced inline). Same behaviour with a semicolon.

julia> foo(wrongkwarg=7, 3, #TAB  -> nothing

julia> foo(1; #TAB  -> nothing

julia> identity(x, unknown..., #TAB
identity(x) in Base at operators.jl:526
```
whereas the current state is:
```julia
julia> foo(a::Int, b, args...; kwarg) = nothing;

julia> foo(3, 4, 5, 6; kw #TAB    -> nothing, no keyword argument completion

julia> foo(wrongkwarg=7, 3, #TAB
foo(a::Int64, b, args...; kwarg) in Main at REPL[1]:1  # should not appear because of the incompatible kwarg

julia> foo(1; #TAB
foo(a::Int64, b, args...; kwarg) in Main at REPL[1]:1  # should not appear because foo requires at least two arguments

julia> identity(x, unknown..., #TAB    -> nothing, but the method should be provided in case unknown is empty
```

CI failure looks unrelated. Could I bother you @aviatesk for another review?

aviatesk commented on stdlib/REPL/test/replcompletions.jl:
```diff
@@ -537,27 +547,56 @@ let s = "CompletionFoo.kwtest( "
     @test length(c) == 1
-    @test occursin("x, y, w...", c[1])
+    @test occursin("a; x, y, w...", c[1])
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(1; x=1, ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(1, x=1; ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(1, kw=1, ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(1; kw=1, ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(1, kw=1; ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(y=3, 1, ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(y=3, 1; ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(kw=3, 1, ")
+    @test (c, r, res) == test_complete("CompletionFoo.kwtest2(kw=3, 1; ")
```
Why shouldn't these example be completed?
For example, `CompletionFoo.kwtest2(1; |` is better provide keyword completions?

Liozou replied:
This is testing that the completion is the same for all these inputs, so it is indeed checking the these examples are completed. The completion is not empty since we're checking just above that `length(c) == 1`.
I added two new lines
```julia
@test (c, r, res) == test_complete("CompletionFoo.kwtest2(1; )
@test (c, r, res) == test_complete("CompletionFoo.kwtest2(1, )
```
since you're right, the completion should also be the same without any given kwarg.

Liozou commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -523,45 +532,50 @@ function complete_methods(ex_org::Expr, context_module::Module=Main)
     seen = Base.IdSet()
     for name in names(callee_module; all=true)
-        if !Base.isdeprecated(callee_module, name) && isdefined(callee_module, name)
+        if !Base.isdeprecated(callee_module, name) && isdefined(callee_module, name) && !startswith(string(name), "#kw")
```
Oh you're absolutely right, I wanted to remove internal functions like kwsorters and such, but in the end this does not make sense as you point out. It can either be replaced by `!startswith(string(name), '#')`, or we can drop that condition altogether and let internal functions appear. I don't think internal functions should appear here (if you want to access them, REPL completion is probably not the best entry point) but I don't mind dropping the condition if you think otherwise?

Liozou commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -583,42 +597,117 @@ function complete_any_methods(ex_org::Expr, callee_module::Module, context_modul
-    t_in = Tuple{funct, args_ex...}
+    num_splat = 0 # number of splat arguments in args_ex
+    args_ex_onevararg = copy(args_ex) # args_ex_onevararg contains at most one Vararg, put in final position
+    for (i, arg) in enumerate(args_ex)
```
Sure, I suppose the compiler prefers this since the type of `arg` cannot be inferred here, is that right?

Liozou commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -583,42 +597,117 @@ function complete_any_methods(ex_org::Expr, callee_module::Module, context_modul
+                    n = isexpr(x, :kw) ? x.args[1] : x
+                    if !(n isa Symbol) || !isexpr(x, :...)
+                        kwargs_flag = 2
+                    end
```
I ended up refactoring `complete_methods_args` as well since the kwarg logic was entirely shared between the two branches (with and without broadcasting)

Liozou commented:
Thanks a lot for the review @aviatesk! https://github.com/JuliaLang/julia/pull/43536/commits/45eccadf42115ca4d56dd4a2d9c35d0555f62567 contains the revisions.

vtjnash commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -583,42 +597,105 @@ function complete_any_methods(ex_org::Expr, callee_module::Module, context_modul
+    return 2 # The kwarg is invalid
+end
+
+function detect_args_kwargs(funargs::Vector{Any}, context_module::Module, default_any::Bool, ::Val{broadcasting}) where {broadcasting}
```
This never should be a `Val`. That just makes the API smelly and performance bad for no reason.
```suggestion
function detect_args_kwargs(funargs::Vector{Any}, context_module::Module, default_any::Bool, broadcasting::Bool)
```

vtjnash commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -583,42 +597,105 @@ function complete_any_methods(ex_org::Expr, callee_module::Module, context_modul
     m isa Vector || return
     for match in m
-        # TODO: if kwargs_ex, filter out methods without kwargs?
+        if exact_narg
+            # If there is a semicolon, the number of non-keyword arguments in the
+            # call cannot grow. It must thus match that of the method.
+            meth_nargs = match.method.nargs - 1 # remove the function itself
+            exn_args = length(args_ex)
+            if meth_nargs != exn_args
+                # the match may still hold if some arguments are slurped or splat
+                num_slurp = count(Base.isvarargtype, Base.unwrap_unionall(match.method.sig).types)
+                num_slurp == 0 && num_splat == 0 && continue
+                if num_slurp == 0
+                    meth_nargs ≥ exn_args - num_splat || continue
+                elseif num_splat == 0
+                    exn_args ≥ meth_nargs - num_slurp || continue
+                end
+            end
+        end
+
```
This is not how we define these values (it is stored in `match.method.isva`), and it won't compute it reliably. But we already computed this when creating `args_ex`, when deciding whether to append `Vararg{Any}`, so none of this code seems necessary. Otherwise the list you are filtering would already be wrong, and it would already be hopeless to fix.

```suggestion
```

vtjnash replied:
This is not how we define these values (it is stored in `match.method.isva`), and it won't compute it reliably. But we already computed this when creating `args_ex`, when deciding whether to append `Vararg{Any}`, so none of this code seems necessary. Otherwise the list you are filtering would already be wrong, and it would already be hopeless to fix.

```suggestion
```

vtjnash commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -732,6 +809,71 @@ end
+function complete_keyword_argument(partial, last_idx, context_module)
+    fail = Completion[], 0:-1
+
+    # Quickly abandon if the situation does not look like the completion of a kwarg
+    idx_last_punct = something(findprev(x -> ispunct(x) && x != '_' && x != '!', partial, last_idx), 0)::Int
+    idx_last_punct == 0 && return fail
+    last_punct = partial[idx_last_punct]
+    last_punct == ',' || last_punct == ';' || last_punct == '(' || return fail
+    before_last_word_start = something(findprev(in(non_identifier_chars), partial, last_idx), 0)
+    before_last_word_start == 0 && return fail
+    all(isspace, partial[nextind(partial, idx_last_punct):before_last_word_start]) || return fail
+    frange, method_name_end = find_start_brace(partial[1:idx_last_punct])
+    method_name_end ∈ frange || return fail
+
+    # At this point, we are guaranteed to be in a set of parentheses, possibly a function
+    # call, and the last word (currently being completed) has no internal dot (i.e. of the
+    # form "foo.bar") and is directly preceded by `last_punct` (one of ',', ';' or '(').
+    # Now, check that we are indeed in a function call
+    frange = first(frange):(last_punct==';' ? prevind(partial, idx_last_punct) : idx_last_punct)
+    s = replace(partial[frange], r"\!+([^=\(]+)" => s"\1") # strip preceding ! operator
+    ex = Meta.parse(s * ')', raise=false, depwarn=false)
+    isa(ex, Expr) || return fail
+    ex.head === :call || (ex.head === :. && ex.args[2] isa Expr && (ex.args[2]::Expr).head === :tuple) || return fail
```
Could we possibly combine much of this work with `complete_methods`? Seems like some at least of this code must be duplicated and extraneous.

Liozou commented:
Thanks a lot @vtjnash!
I remember there were some subtleties that motivated the use the `if exact_narg` block in addition to the previous check for `Vararg`, I think it was to disallow the completion of `sum(somesplat..., too, many, arguments, afterwards)` or something like that. But anyway, I'll update it to use `match.method.isva` (which I wasn't aware of, thanks!), and I'll try to make it cleaner if possible. I'll look into factoring some of the code with `complete_methods` as well (and fix the `Val` thing).
I'm not sure when I have time to do this though, but hopefully by next week.

Liozou commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -732,6 +809,71 @@ end
+    kwargs = Set{String}()
+    for m in methods
+        m::MethodCompletion
+        possible_kwargs = Base.kwarg_decl(m.method)
```
This can apparently crash in some cases (only builtins maybe?). For example:
```julia
julia> isdefined(res┌ Error: Error in the keymap
│   exception =
│    TypeError: in typeassert, expected Core.MethodTable, got a value of type Nothing
│    Stacktrace:
│      [1] get_methodtable
│        @ ./reflection.jl:1657 [inlined]
│      [2] kwarg_decl(m::Method, kwtype::Nothing) (repeats 2 times)
│        @ Base ./methodshow.jl:82
│      [3] complete_keyword_argument(partial::String, last_idx::Int64, context_module::Module)
│        @ REPL.REPLCompletions ~/julia/usr/share/julia/stdlib/v1.9/REPL/src/REPLCompletions.jl:861
...
```

Moelf commented:
what's the status of this PR?

HiperMaximus commented:
> what's the status of this PR?

Very interested in this PR

Liozou commented:
As mentioned in [the comment before](https://github.com/JuliaLang/julia/pull/43536#issuecomment-1058615091), I believe https://github.com/JuliaLang/julia/pull/44434 is a prerequisite for this because many kwarg completions will unexpectedly fail otherwise, which is arguably worse than having no completion at all. So I'm still waiting on that PR to be discussed before updating this one (with both a rebase on master and taking into account the review comments above).

vtjnash commented:
to decouple those 2 PR efforts, I think here we might be better to treat any `kwargs...` as having the union of all possible explicit keywords of `method(test)`, just for completion purposes. So ignore which method each applies to, and just list all of them

Liozou committed:
Add REPL-completion for keyword arguments

Liozou commented:
Rebase done! So, I removed all the stuff that prevented kwarg completion when the call signature did not correspond to the method, because it would lead to false-negatives (i.e. some completions missing, see [comment above](https://github.com/JuliaLang/julia/pull/43536#issuecomment-1058615091)). As a consequence, we now have false-positives (i.e. some completions provided when they should not) but no false-negatives, for example:
```julia
julia> foo(a::Int, bar=12, qux=""; kwarg) = nothing
foo (generic function with 3 methods)

julia> foo(387; kwa#TAB
julia> foo(387; kwarg=  # expected completion, great!

julia> foo("a"; kwa#TAB
julia> foo("a"; kwarg=  # false-positive, this should not complete.
```

For safe-keeping, the previous state of the PR is available at https://github.com/JuliaLang/julia/compare/master...Liozou:julia:replcompletekwargs_save, including additional handling of splats that should come in a separate PR. I added the tests that passed there but not here as comments with a "TODO:" instead of `@test_broken` because there are wishful features, not bugs really. I do plan on fixing them if/when https://github.com/JuliaLang/julia/pull/44434 is accepted.

vtjnash commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -744,6 +770,76 @@ end
+function identify_possible_method_completion(partial, last_idx)
+    fail = 0:-1, Expr(:nothing), 0:-1, 0
+
+    # First, check that the last punctuation is either ',', ';' or '('
+    idx_last_punct = something(findprev(x -> ispunct(x) && x != '_' && x != '!', partial, last_idx), 0)::Int
+    idx_last_punct == 0 && return fail
+    last_punct = partial[idx_last_punct]
+    last_punct == ',' || last_punct == ';' || last_punct == '(' || return fail
+
+    # Then, check that `last_punct` is only followed by an identifier or nothing
+    before_last_word_start = something(findprev(in(non_identifier_chars), partial, last_idx), 0)
+    before_last_word_start == 0 && return fail
+    all(isspace, @view partial[nextind(partial, idx_last_punct):before_last_word_start]) || return fail
+
+    # Check that `last_punct` is either the last '(' or placed after a previous '('
+    frange, method_name_end = find_start_brace(@view partial[1:idx_last_punct])
+    method_name_end ∈ frange || return fail
+
+    # Strip the preceding ! operators, if any, and close the expression with a ')'
+    s = replace(partial[frange], r"\G\!+([^=\(]+)" => s"\1"; count=1) * ')'
+    ex = Meta.parse(s, raise=false, depwarn=false)
+    isa(ex, Expr) || return fail
```
Seems unfortunate we need all this code, since it seems like a slightly less complete version of https://github.com/JuliaLang/julia/blob/9bdaabdc829f7d26d8ab34cf422612161777f870/stdlib/REPL/src/REPLCompletions.jl#L905-L926
followed by a copy of
https://github.com/JuliaLang/julia/blob/9bdaabdc829f7d26d8ab34cf422612161777f870/stdlib/REPL/src/REPLCompletions.jl#L831-L834

vtjnash replied:
okay, I see the purpose is a bit different, since we are just looking for whether there is something quite trivial after the last punctuation, and the other code (and should_method_complete) is moved here now

vtjnash commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -744,6 +770,76 @@ end
+    kwargs_flag == 2 && return fail # one of the previous kwargs is invalid
+
+    methods = Completion[]
+    complete_methods!(methods, funct, Any[Vararg{Any}], kwargs_ex, -1, kwargs_flag == 1)
+    # TODO: use args_ex instead of Any[Vararg{Any}] and only provide kwarg completion for
+    # method calls compatible with the current arguments.
```
In theory, this is a pretty bad choice of methods to consider, since it is specifically only looking at methods without kwargs. And you never want to pass `-1` here, as you can accidentally break interactivity for several minutes. So without those bad trailing args, and looking only at methods with kwargs, we want something like this:

```julia
# get the list of possible kw method table
kwt = try; Core.kwftype(funct); catch ex; ex isa ErrorException || rethrow(); return fail; end
ml = Completion[]
complete_methods!(ml, kwt, Any[Any, funct, args_ex...])
ml = Method[m.method for m in ml if m isa MethodCompletion]
# decide if we might have a `...` in the list
have_kwarg = any(isa(TextCompletion), ml)
for m in ml
    kwli = m.method
    slotnames = ccall(:jl_uncompress_argnames, Vector{Symbol}, (Any,), kwli.slot_syms)
    have_kwarg = any(endswith("..."), slotnames)
    have_kwarg && break
end
if have_kwarg
    # if we have a `kwarg...`, instead return the union of all possible method kwargs for this function
    methods = Base.MethodList(kwt.name.mt)
end
# now get the kwarg names, and filter them to just the matching ones
kwargs = Set{String}()
for m in ml
    kwli = m.method
    slotnames = ccall(:jl_uncompress_argnames, Vector{Symbol}, (Any,), kwli.slot_syms)
    union!(kwargs, filter!(slotnames))
end
```

(kwargs_ex should not be passed here either, as that is sort of just an historical accident of past implementation problems)

pbouffard commented:
I just stumbled across this in 1.9.0-beta4, awesome. But I don't see it in the docs for REPL tab completion nor in the NEWS file. This is a great feature addition which should be advertised! :)
